### PR TITLE
Remove synchronized blocks in Thread isAlive and isDead methods

### PR DESCRIPTION
**Pull request by babsingh**
Made `threadRef` volatile so that the most up-to-date value is seen
by all the threads.

`threadRef = NO_REF` assignment will happen once, and similarly, the
`started` field is also set once during initialization.

A user of these methods will probably invoke them multiple times
until the desired result is achieved. A delay in getting the most
up-to-date value should not hinder the functionality of these
methods.

To sum up, removing synchronization in these methods improves perf
by reducing lock contention without hindering the functionality.

Reference to `PR 1FJMO7Q` has been removed and a comment related to
`ThreadGroup` has been updated since the code seems to have evolved
since the comments were written.

Related: #20414

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
jenkins test sanity amac jdk17

**Comment by babsingh** [comment]
Updated a typo in the commit message. The previously launched PR build should still be good: https://openj9-jenkins.osuosl.org/view/Pull%20Requests/job/PullRequest-OpenJ9/6285.

File: jcl/src/java.base/share/classes/java/lang/Thread.java [hunk 1]
```diff
@@ -746,10 +746,8 @@ public static boolean interrupted() {
  * @see			Thread#start
  */
 public final boolean isAlive() {
-	synchronized (lock) {
```
Comment (ThanHenderson) [inline jcl/src/java.base/share/classes/java/lang/Thread.java:749]: Do we have the source as to why these locks were acquired here in the first place? I suspect it has something to do with guarding the `volatile` variables.

File: jcl/src/java.base/share/classes/java/lang/Thread.java [hunk 2]
```diff
@@ -765,9 +763,7 @@ public final boolean isAlive() {
  */
 private boolean isDead() {
 	// Has already started, is not alive anymore, and has been removed from the ThreadGroup
```
Comment (ThanHenderson) [inline jcl/src/java.base/share/classes/java/lang/Thread.java:765]: This comment (and the doc comment) suggests we are doing some type of checking w.r.t. removal from a ThreadGroup. Do we know if `started` indicates anything about ThreadGroup removal? Either way, while we are here we should ensure that the comment is accurate.

File: jcl/src/java.base/share/classes/java/lang/Thread.java (hunk 1 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/Thread.java:749]: They seem to exist since the beginning as means to protect access to shared variables. I didn't find a special reason for their use here. Based on how on how these methods will be used, they should function fine without synchronization.

Comment (ThanHenderson) [reply jcl/src/java.base/share/classes/java/lang/Thread.java:749]: I also checked in the extension repo code for j.l.Thread that we use for the more recent JDK versions, and `isAlive` already doesn't have the `synchronized` block.`isDead` still does though. Are you opening a patch for the extension repos as well to remove the `isDead` lock?

File: jcl/src/java.base/share/classes/java/lang/Thread.java (hunk 2 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/Thread.java:765]: `threadRef` can be NULL before the thread starts and after it exits. `started = TRUE` makes sure that we only pick up the exit case and implies that thread has been removed from the `ThreadGroup`.

Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/Thread.java:765]: What will be your suggestion for a rewrite?

Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/Thread.java:765]: There is also a reference to a very old issue: PR 1FJMO7Q; I believe it is no longer accessible. They might have seen an issue related to `ThreadGroup` and `isDead` method.

May be, change it to the following to clarify and highlight the implicit relationship?
```suggestion
	/* threadRef can be NO_REF before the thread starts and after it exits.
	 * Only account for the exit case where threadRef is NO_REF. This assures
	 * that the thread has been removed from the ThreadGroup.
	 */ 
```

File: jcl/src/java.base/share/classes/java/lang/Thread.java (hunk 1 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/Thread.java:749]: > Are you opening a patch for the extension repos as well to remove the isDead lock?

Yes.

File: jcl/src/java.base/share/classes/java/lang/Thread.java (hunk 2 above)
Comment (ThanHenderson) [reply jcl/src/java.base/share/classes/java/lang/Thread.java:765]: > implies that thread has been removed from the ThreadGroup.

It was this implication that isn't clear to me. I'd've thought that:
 `// Had already started and is not alive anymore.`
 would alone be appropriate as the comment.
 
The only place I could find where the thread is removed from the ThreadGroup is [here](https://github.com/eclipse-openj9/openj9/blob/6544c13e4e911528290f3d546024780f307e6e0e/jcl/src/java.base/share/classes/java/lang/Thread.java#L1150).

But that isn't through the `exit()` path. That is in the failure to `start()` path. Exit simply `null`s out the `group` member.

Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/Thread.java:765]: Oh, I see. It looks like the code has changed; but, the comment was never updated. We can probably remove the reference to the old issue and `ThreadGroup`.

Comment (ThanHenderson) [reply jcl/src/java.base/share/classes/java/lang/Thread.java:765]: That is what I am thinking too. But I am also wondering now if we are missing a call to `this.group.remove(this)` during the exit sequence.

File: jcl/src/java.base/share/classes/java/lang/Thread.java [hunk 3]
```diff
@@ -764,10 +761,8 @@ public final boolean isAlive() {
  * @see			Thread#isAlive
  */
 private boolean isDead() {
-	// Has already started, is not alive anymore, and has been removed from the ThreadGroup
-	synchronized(lock) {
-		return started && threadRef == NO_REF;
-	}
+	/* Has already started and is not alive anymore. */
+	return started && threadRef == NO_REF;
```
Comment (ThanHenderson) [inline jcl/src/java.base/share/classes/java/lang/Thread.java:765]: suggest for readability and match with the extension repo versions:
`return (started && (threadRef == NO_REF))`

File: jcl/src/java.base/share/classes/java/lang/Thread.java (hunk 2 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/Thread.java]: In OpenJ9's `ThreadGroup.java` (JDK17-), there is a comment which is repeated twice: `Dead Threads not removed from ThreadGroups`. There might have been a behavior change where dead threads were no longer removed from the `ThreadGroup`. Also, methods like `ThreadGroup.activeCount` are anticipating dead threads since they are doing checks like `isAlive`.

In JDK21+, `ThreadGroup` in the extension repo no longer maintains a list of threads; instead, it relies upon `Thread.getAllThreads()`.

`ThreadGroups` are not widely used and considered unsafe ([ref](https://stackoverflow.com/questions/18897621/why-is-not-safe-to-use-java-lang-threadgroup)). Since no functional issues are reported, it might not be worth investigating this issue.

Comment (ThanHenderson) [reply jcl/src/java.base/share/classes/java/lang/Thread.java]: Sounds good.

**Comment by ThanHenderson** [comment]
lgtm

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk17

**Comment by tajila** [comment]
jenkins test extended.system zlinux64 jdk17

**Comment by tajila** [comment]
jenkins test extended.functional plinux64 jdk17

**Comment by tajila** [comment]
jenkins test extended.system zlinux jdk17

**Comment by tajila** [comment]
jenkins test extended.functional plinux jdk17

**Comment by tajila** [comment]
Actually, it was probably `getStackTrace` and the thread dying. Anyways, Ill see if I can find the test.

**Comment by ThanHenderson** [comment]
The other uses of locks that wrap `isAlive()` weren't removed e.g. https://github.com/eclipse-openj9/openj9/blob/c0c04fc18302811c3330f4ea2ef0ce5f2a825b95/jcl/src/java.base/share/classes/java/lang/Thread.java#L1345-L1351
So it's likely that the test you are referring to wouldn't regress.

**Comment by tajila** [comment]
I found the test, it was an issue with throwing an exception while the thread was dying, so its unaffected byt theses changes.

**Review by tajila — APPROVED** [review]
